// devextreme/esm/renovation/ui/toolbar/toolbar_props.js
/**
 * DevExtreme (esm/renovation/ui/toolbar/toolbar_props.js)
 * Version: 24.1.6
 * Build date: Sat Sep 14 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    BaseWidgetProps
} from "../common/base_props";
export const CollectionWidgetItem = {};
export const BaseToolbarItemProps = {};
export const ToolbarTextBoxProps = Object.create(Object.prototype, Object.assign(Object.getOwnPropertyDescriptors(BaseToolbarItemProps), Object.getOwnPropertyDescriptors({
    value: ""
})));
export const ToolbarCheckBoxProps = Object.create(Object.prototype, Object.assign(Object.getOwnPropertyDescriptors(BaseToolbarItemProps), Object.getOwnPropertyDescriptors({
    value: false
})));
export const ToolbarButtonGroupProps = BaseToolbarItemProps;
export const ToolbarButtonProps = BaseToolbarItemProps;
export const ToolbarDropDownButtonItemProps = CollectionWidgetItem;
export const ToolbarDropDownButtonItemPropsType = {};
export const ToolbarDropDownButtonProps = BaseToolbarItemProps;
export const ToolbarItem = CollectionWidgetItem;
export const ToolbarItemType = {};
export const ToolbarProps = BaseWidgetProps;
