// devextreme/cjs/__internal/scheduler/r1/appointment_popup/index.js
/**
 * DevExtreme (cjs/__internal/scheduler/r1/appointment_popup/index.js)
 * Version: 24.1.6
 * Build date: Sat Sep 14 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "POPUP_WIDTH", {
    enumerable: true,
    get: function() {
        return _config.POPUP_WIDTH
    }
});
Object.defineProperty(exports, "defaultAnimation", {
    enumerable: true,
    get: function() {
        return _config.defaultAnimation
    }
});
Object.defineProperty(exports, "getMaxWidth", {
    enumerable: true,
    get: function() {
        return _config.getMaxWidth
    }
});
Object.defineProperty(exports, "getPopupSize", {
    enumerable: true,
    get: function() {
        return _config.getPopupSize
    }
});
Object.defineProperty(exports, "getPopupToolbarItems", {
    enumerable: true,
    get: function() {
        return _config.getPopupToolbarItems
    }
});
Object.defineProperty(exports, "isPopupFullScreenNeeded", {
    enumerable: true,
    get: function() {
        return _config.isPopupFullScreenNeeded
    }
});
var _config = require("./config");
